src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -11,46 +11,75 @@
 
 namespace Microsoft.AspNet.OData.Extensions
 {
-    internal static class ODataPathExtensions
+    /// <summary>
+    /// Extensions method for <see cref="ODataPath"/>.
+    /// </summary>
+    public static class ODataPathExtensions
     {
+        /// <summary>
+        /// Get keys from the last <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path"><see cref="ODataPath"/>.</param>
+        /// <returns>Dictionary of keys.</returns>
         public static Dictionary<string, object> GetKeys(this ODataPath path)
         {
             if (path == null)
             {
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
+            int position = path.Count - 1;
+            ODataPathSegment pathSegment = pathSegments[position];
```

Comment:
Can an OData path have 0 segments? That would cause us problems here

Reply:
Resolved